openmc/lattice.py

```diff
@@ -0,0 +1,1901 @@
+        of the lattice
+
+    """
+
+    next_id = 1
+    used_ids = openmc.Universe.used_ids
+
+    def __init__(self, lattice_id=None, name=''):
+        # Initialize Lattice class attributes
+        self.id = lattice_id
+        self.name = name
+        self._pitch = None
+        self._outer = None
+        self._universes = None
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def pitch(self):
+        return self._pitch
+
+    @property
+    def outer(self):
+        return self._outer
+
+    @property
+    def universes(self):
+        return self._universes
+
+    @name.setter
+    def name(self, name):
+        if name is not None:
+            cv.check_type('lattice name', name, str)
+            self._name = name
+        else:
+            self._name = ''
+
+    @outer.setter
+    def outer(self, outer):
+        cv.check_type('outer universe', outer, openmc.Universe)
+        self._outer = outer
+
+    @staticmethod
+    def from_hdf5(group, universes):
+        """Create lattice from HDF5 group
+
+        Parameters
+        ----------
+        group : h5py.Group
+            Group in HDF5 file
+        universes : dict
+            Dictionary mapping universe IDs to instances of
+            :class:`openmc.Universe`.
+
+        Returns
+        -------
+        openmc.Lattice
+            Instance of lattice subclass
+
+        """
+        lattice_id = int(group.name.split('/')[-1].lstrip('lattice '))
+        name = group['name'][()].decode() if 'name' in group else ''
+        lattice_type = group['type'][()].decode()
+
+        if lattice_type == 'rectangular':
+            dimension = group['dimension'][...]
+            lower_left = group['lower_left'][...]
+            pitch = group['pitch'][...]
+            outer = group['outer'][()]
+            universe_ids = group['universes'][...]
+
+            # Create the Lattice
+            lattice = openmc.RectLattice(lattice_id, name)
+            lattice.lower_left = lower_left
+            lattice.pitch = pitch
+
+            # If the Universe specified outer the Lattice is not void
+            if outer >= 0:
+                lattice.outer = universes[outer]
+
+            # Build array of Universe pointers for the Lattice
+            uarray = np.empty(universe_ids.shape, dtype=openmc.Universe)
+
+            for z in range(universe_ids.shape[0]):
+                for y in range(universe_ids.shape[1]):
+                    for x in range(universe_ids.shape[2]):
+                        uarray[z, y, x] = universes[universe_ids[z, y, x]]
+
+            # Use 2D NumPy array to store lattice universes for 2D lattices
+            if len(dimension) == 2:
+                uarray = np.squeeze(uarray)
+                uarray = np.atleast_2d(uarray)
+
+            # Set the universes for the lattice
+            lattice.universes = uarray
+
+        elif lattice_type == 'hexagonal':
```

Comment:
Is there any other possible lattice type? Even if not -- a default `else` case should be added that throws an appropriate error